### PR TITLE
feat(cast): add list support for upgrade result in cast

### PR DESCRIPTION
**Pull request by sagarkrsd**
**What this PR does / why we need it**:

This PR adds the list and get support in cast for upgrade-result CR as per new folder structure for adding supports to runtasks.

Sample runtask to get UpgradeResult CR:
```yaml
apiVersion: openebs.io/v1alpha1
kind: RunTask
metadata:
  name: get-ur-info
  namespace: default
spec:
  meta: |
    id: geturinfo
    apiVersion: openebs.io/v1alpha1
    kind: UpgradeResult
    action: get
    objectName: vol-upgrade-result
  post: |-
    '{{- jsonpath .JsonResult "{.status.failedCount}" | trim | saveAs "failed_count" .TaskResult | noop -}}'
```

Sample runtask to list UpgradeResult CR:
```yaml
apiVersion: openebs.io/v1alpha1
kind: RunTask
metadata:
  name: list-upgrade-results
  namespace: default
spec:
  meta: |
    id: listUpgradeResults
    apiVersion: openebs.io/v1alpha1
    kind: UpgradeResult
    action: list
    runNamespace: default
  post: |-
    '{{- jsonpath .JsonResult "{range .items[*]}{.metadata.name} {end}" | trim | saveAs "upgrade_results" .TaskResult | noop -}}'
```

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

Signed-off-by: sagarkrsd <sagar.kumar@openebs.io>

### CODE REVIEW

File: pkg/upgrade/v1alpha1/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	cs, err := k.getClientOrCached()
+	if err != nil {
+		return nil, err
+	}
+	return k.list(cs, namespace, opts)
+}
+
+// Get returns an upgrade result instance from kubernetes cluster
+func (k *kubeclient) Get(name string, opts metav1.GetOptions) (*apis.UpgradeResult, error) {
```
Comment (sagarkrsd) [inline pkg/upgrade/v1alpha1/kubernetes.go]: This is still not fully added. Will add all the supports once the approach gets finalized.

File: pkg/upgrade/v1alpha1/upgraderesult.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
+)
+
+type upgradeResult struct {
+	// upgrade result object
+	object apis.UpgradeResult
+}
+
+type urList struct {
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/upgraderesult.go]: let this be named as `upgradeResultList`

File: pkg/upgrade/v1alpha1/upgraderesult.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+package v1alpha1
+
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
+)
+
+type upgradeResult struct {
+	// upgrade result object
+	object apis.UpgradeResult
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/upgraderesult.go]: use pointer. i.e. `object   *apis.UpgradeResult`

File: pkg/upgrade/v1alpha1/upgraderesult.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+type urList struct {
+	// list of upgrade results
+	items []upgradeResult
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/upgraderesult.go]: use pointer `items   []*upgradeResult`

File: pkg/task/task.go [hunk 5]
```diff
@@ -28,12 +29,14 @@ import (
...
 	podexec "github.com/openebs/maya/pkg/kubernetes/podexec/v1alpha1"
 	"github.com/openebs/maya/pkg/template"
+	upgrade "github.com/openebs/maya/pkg/upgrade/v1alpha1"
 	"github.com/openebs/maya/pkg/util"
 	api_apps_v1 "k8s.io/api/apps/v1"
 	api_apps_v1beta1 "k8s.io/api/apps/v1beta1"
 	api_batch_v1 "k8s.io/api/batch/v1"
 	api_core_v1 "k8s.io/api/core/v1"
 	api_extn_v1beta1 "k8s.io/api/extensions/v1beta1"
+	mach_apis_meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (AmitKumarDas) [inline pkg/task/task.go]: can we alias this to `metav1`

File: pkg/upgrade/v1alpha1/kubernetes.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+			config, err := client.GetConfig(k.getClient())
+			if err != nil {
+				return nil, err
+			}
+			return clientset.NewForConfig(config)
+		}
+	}
+	if k.list == nil {
+		k.list = func(cs *clientset.Clientset, opts metav1.ListOptions) (*apis.UpgradeResultList, error) {
+			return cs.OpenebsV1alpha1().UpgradeResults(k.namespace).List(opts)
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/kubernetes.go]: I am not sure if using `k.namespace` will workout?
If yes, then we can think of making use of `k.clientset` as well.
We need to do UT to understand the implications.

File: pkg/upgrade/v1alpha1/kubernetes.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+type listFunc func(cs *clientset.Clientset, opts metav1.ListOptions) (*apis.UpgradeResultList, error)
+
+// getFunc is a typed function that abstracts
+// getting upgrade result instances
+type getFunc func(cs *clientset.Clientset, name string, opts metav1.GetOptions) (*apis.UpgradeResult, error)
+
+// kubeclient enables kubernetes API operations
+// on upgrade result instance
+type kubeclient struct {
+	client    *client.Client
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/kubernetes.go]: we may not need to have a client & getClient

File: pkg/upgrade/v1alpha1/kubernetes.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	if k.list == nil {
+		k.list = func(cs *clientset.Clientset, namespace string, opts metav1.ListOptions) (*apis.UpgradeResultList, error) {
+			return cs.OpenebsV1alpha1().UpgradeResults(namespace).List(opts)
+		}
+	}
+}
+
+// WithClientset sets the kubernetes client against
+// the kubeclient instance
+func WithClientset(c *clientset.Clientset) kubeclientBuildOption {
```
[the remark this answers is not in the document]
Comment (AmitKumarDas) [reply pkg/upgrade/v1alpha1/kubernetes.go]: If we dont need client & then WithClient will not be required

File: pkg/upgrade/v1alpha1/kubernetes.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
+/*
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/kubernetes.go]: we might want to package this into `pkg/upgrade/result/v1alpha1/`

File: pkg/upgrade/v1alpha1/upgraderesult.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/upgraderesult.go]: we might need to package this under `pkg/upgrade/result/v1alpha1`

File: pkg/upgrade/v1alpha1/upgraderesult.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+// WithAPIList builds the list of ur
+// instances based on the provided
+// ur api instances
+func (b *listBuilder) WithAPIList(list *apis.UpgradeResultList) *listBuilder {
+	if list == nil {
+		return b
+	}
+	for _, c := range list.Items {
+		b.list.items = append(b.list.items, &upgradeResult{object: &c})
```
Comment (AmitKumarDas) [inline pkg/upgrade/v1alpha1/upgraderesult.go]: check if this works. perhaps UT.
we might need:
```go
for i, _ := range list.Items {
  b.list.items = append(b.list.items, &upgradeResult{object: &list.Items[i]}
}
```

File: pkg/upgrade/result/v1alpha1/kubernetes_test.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+			}
+			if !mock.expectNamespaceEmpty && fake.namespace == "" {
+				t.Fatalf("test %q failed expected fake.namespace not to be empty", name)
+			}
+		})
+	}
+}
+func TestGetClientOrCached(t *testing.T) {
+	tests := map[string]struct {
+		expectErr  bool
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes_test.go]: can we rename to `isError`

File: pkg/upgrade/result/v1alpha1/kubernetes_test.go [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+		KubeClient *kubeclient
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes_test.go]: Can we make this private i.e. `kubeClient`

File: pkg/upgrade/result/v1alpha1/kubernetes_test.go [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+				t.Fatalf("test %q failed : expected clientset %v but got %v", name, mock.KubeClient.clientset, c)
+			}
+		})
+	}
+}
+func TestKubernetesList(t *testing.T) {
+	tests := map[string]struct {
+		getClientset getClientsetFunc
+		list         listFunc
+		expectErr    bool
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes_test.go]: can we rename to `isError`

File: pkg/upgrade/result/v1alpha1/upgraderesult.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+	if list == nil {
+		return b
+	}
+	for i := range list.Items {
+		b.list.items = append(b.list.items, &upgradeResult{object: &list.Items[i]})
+	}
+	return b
+}
+
+// List returns the list of ur
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/upgraderesult.go]: let us not use short forms i.e. `ur`

File: pkg/upgrade/result/v1alpha1/upgraderesult_test.go [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/upgrade/v1alpha1"
+)
+
+func TestWithAPIList(t *testing.T) {
+	tests := map[string]struct {
+		expectedURName []string
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/upgraderesult_test.go]: A UT struct should be always designed with `inputs` & `expected outputs` in mind.
Can you redesign this accordingly?
It is okay to provide the same arguments twice one as input & other as expected output in UT.
In this case, we can even check the len of inputs as an expectation i.e. `expectedLen int`

File: pkg/upgrade/result/v1alpha1/kubernetes_test.go [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+		"Namespace is not empty": {"abc", false},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			h := WithNamespace(mock.namespace)
+			fake := &kubeclient{}
+			h(fake)
+			if mock.expectNamespaceEmpty && fake.namespace != "" {
+				t.Fatalf("test %q failed expected fake.namespace to be empty", name)
```
[the remark this answers is not in the document]
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes_test.go]: Was using uid first then changed it to use name, will change it to %s. Thanks.

**Pull request merged**